### PR TITLE
feat: exclude region for StudyLocus object

### PR DESCRIPTION
**Pull request by Daniel-Considine**
## 🛠 What does this PR implement

Replicated the `exclude_region()` function from `SummaryStatistics` into `StudyLocus`. Additionally, has an option to exclude any StudyLocus where the window, by locusStart and locusEnd, overlaps with the input region. 
## 🚦 Before submitting

- [x] Do these changes cover one single feature (one change at a time)?
- [x] Did you read the [contributor guideline](https://opentargets.github.io/gentropy/development/contributing/#contributing-checklist)?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you make sure there is no commented out code in this PR?
- [ ] Did you follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standards in PR title and commit messages?
- [ ] Did you make sure the branch is up-to-date with the `dev` branch?
- [ ] Did you write any new necessary tests?
- [ ] Did you make sure the changes pass local tests (`make test`)?
- [ ] Did you make sure the changes pass pre-commit rules (e.g `poetry run pre-commit run --all-files`)?

### CODE REVIEW

**Comment by addramir** [comment]
Question about overlap logic:
I thought that we should exclude all regions that have overlap with boundaries, not entirely within the specified region.
Something like this:
```
if exclude_overlap:
    filter_condition = ~(
        (f.col("chromosome") == chromosome)
        & (
            ((f.col("locusStart") <= end_position)
            & (f.col("locusStart") >= start_position)) |
            ((f.col("locusEnd") <= end_position)
            & (f.col("locusEnd") >= start_position))
        )
    )
```

**Comment by Daniel-Considine** [comment]
@addramir 
The overlap logic should be removing windows having any overlap with the boundaries. 
When locusStart is less than region_end **and** locusEnd is more than region_start, that means there is some amount of overlap. It should be doing the same thing as the logic in your reply.

**Comment by addramir** [comment]
Yes, I see, it should be the same, thanks!

**Review by addramir — APPROVED** [review]

File: src/gentropy/dataset/study_locus.py [hunk 1]
```diff
@@ -574,6 +574,41 @@ def clump(self: StudyLocus) -> StudyLocus:
...
+        Returns:
+            StudyLocus: filtered StudyLocus object.
+        """
+        (chromosome, start_position, end_position) = parse_region(region)
+        if exclude_overlap:
+            filter_condition = ~(
+                (f.col("chromosome") == chromosome)
+                & (
+                    (f.col("locusStart") <= end_position)
+                    & (f.col("locusEnd") >= start_position)
```
Comment (DSuveges) [inline src/gentropy/dataset/study_locus.py:595]: `locusStart` and `locusEnd` columns are nullable. How does this condition works if they are null?

File: src/gentropy/dataset/study_locus.py [hunk 2]
```diff
@@ -574,6 +574,41 @@ def clump(self: StudyLocus) -> StudyLocus:
...
         return self
 
+    def exclude_region(
+        self: StudyLocus, region: str, exclude_overlap: bool = False
+    ) -> StudyLocus:
+        """Exclude a region from the StudyLocus dataset.
+
+        Args:
+            region (str): region given in "chr##:#####-####" format
+            exclude_overlap (bool): If True, excludes StudyLocus windows with any overlap with the region.
```
Comment (DSuveges) [inline src/gentropy/dataset/study_locus.py:584]: > exclude_overlap (bool): If True, excludes StudyLocus windows with any overlap with the region.

I am not 100% sure if this is what happening here. You are excluding all study loci where the lead snp falls within the locus. Not if the start and end position overlap. If I understood it right.

Comment (Daniel-Considine) [reply src/gentropy/dataset/study_locus.py:584]: That's right. By default (exclude_overlap=False) it just looks at the position column, so the position of the lead variant for that StudyLocus. When True, it instead removes any rows where the StudyLocus' window overlaps with the region given to the function.

File: src/gentropy/dataset/study_locus.py (hunk 1 above)
Comment (Daniel-Considine) [reply src/gentropy/dataset/study_locus.py:595]: Just checked - it drops any rows where locusStart/locusEnd are null. `exclude_region()` also drops rows with null for the position column.

File: src/gentropy/dataset/study_locus.py (hunk 2 above)
Comment (DSuveges) [reply src/gentropy/dataset/study_locus.py:584]: So it doesn't care if the actual locusStart / locusEnd overlaps with the filtered region. OK.

**Pull request merged**